### DISCUSSION TITLE
How to populate a dropdown menu using data from an external API?

### DISCUSSION THREAD

**Discussion by MikeAlhayek** [OrchardCMS/OrchardCore#9241] [Q&A]
I am trying to create an attachable content-part that renders a dropdown menu where the options are coming from external API. I am guessing I'll need to create a content-part with a `TextField` field using a `PredefinedList` editor or a content picker of some sort. 

But how can I populate the options from an external API? Do I need to create a custom field for every menu that I want to pull from the API and then attach the new custom field to the part?

**Comment by hishamco** [comment]
I think you need to implement `INavigationProvider` to create an `AdminMenu` that got populated from external APIs, please have a look to and of the `AdminMenu` in the predefine modules

**Reply by MikeAlhayek** [reply]
@hishamco by mean I meant a drop down menu not a navigation menu. I'll update the question for clarification. I want to add a field to a content-part where that field show a drop down menu with the options coming from external API.

**Reply by hishamco** [reply]
I see, I think you could create a custom menu items, have a look to content type selector, it populate the data from the content types, in your case you need to populate them from your source of data

**Reply by MikeAlhayek** [reply]
@hishamco do you mean to use [ContentPickerField](https://docs.orchardcore.net/en/dev/docs/reference/modules/ContentFields/)? If so, do you know where there is a good example to follow?

**Reply by hishamco** [reply]
Yep, content picker

Have a look to [ContentPickerField.cshtml](https://github.com/OrchardCMS/OrchardCore/blob/1ec2e9898ff37e49d4224732c8a3e0acb5ed2fa6/src/OrchardCore.Modules/OrchardCore.ContentFields/Views/ContentPickerField.cshtml) and [ContentPickerFieldDisplayDriver](https://github.com/OrchardCMS/OrchardCore/blob/1ec2e9898ff37e49d4224732c8a3e0acb5ed2fa6/src/OrchardCore.Modules/OrchardCore.ContentFields/Drivers/ContentPickerFieldDisplayDriver.cs)

**Reply by MikeAlhayek** [reply]
@hishamco if I understand correctly, your suggesting to create a custom field for every menu. So if I want to get a drop down menu that has a list of types coming from an API, I would need to create a custom field, and attach it to a custom part. Am I understanding this correctly? Is this the recommended approach?

**Reply by hishamco** [reply]
> @hishamco if I understand correctly, your suggesting to create a custom field for every menu

Could you please explain what kind of menus you are planning to show to understand you, coz my intend not create custom field per menu
